File: src/main/java/dev/nytweetdeck/message/DirectMessageResponseParser.java
package dev.nytweetdeck.message;

import dev.nytweetdeck.message.DirectMessagePage.DirectMessage;
import dev.nytweetdeck.xapi.http.XApiHttpException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.stereotype.Component;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

@Component
public class DirectMessageResponseParser {

    private final ObjectMapper objectMapper;

    public DirectMessageResponseParser(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public DirectMessagePage parse(String body) {
        try {
            var root = objectMapper.readTree(body);
            var state = object(root, "inbox_initial_state");
            if (state == null) {
                state = root;
            }
            var users = parseUsers(object(state, "users"));
            var messages = new ArrayList<DirectMessage>();
            var entries = state.get("entries");
            if (entries != null && entries.isArray()) {
                for (var entry : entries) {
                    var message = object(entry, "message");
                    if (message == null) {
                        continue;
                    }
                    var data = object(message, "message_data");
                    var id = text(message, "id");
                    var senderId = text(data, "sender_id");
                    var messageText = text(data, "text");
                    if (id == null || senderId == null || messageText == null) {
                        continue;
                    }
                    var user = users.getOrDefault(senderId, User.EMPTY);
                    messages.add(new DirectMessage(
                            id,
                            text(message, "conversation_id"),
                            senderId,
                            user.name(),
                            user.username(),
                            user.avatarUrl(),
                            messageText,
                            longValue(message, "time")));
                }
            }
            messages.sort(Comparator.comparingLong(DirectMessage::timestamp).reversed());
            return new DirectMessagePage(
                    messages, firstNonNull(text(state, "cursor"), text(root, "cursor")));
        } catch (JacksonException exception) {
            throw new XApiHttpException("ダイレクトメッセージ応答を解析できません。", exception);
        }
    }

    private static Map<String, User> parseUsers(JsonNode usersNode) {
        var users = new LinkedHashMap<String, User>();
        if (usersNode == null || !usersNode.isObject()) {
            return users;
        }
        for (Map.Entry<String, JsonNode> property : usersNode.properties()) {
            var value = property.getValue();
            users.put(
                    property.getKey(),
                    new User(
                            text(value, "name"),
                            text(value, "screen_name"),
                            text(value, "profile_image_url_https")));
        }
        return users;
    }

    private static JsonNode object(JsonNode node, String field) {
        if (node == null) {
            return null;
        }
        var value = node.get(field);
        return value != null && value.isObject() ? value : null;
    }

    private static String text(JsonNode node, String field) {
        if (node == null) {
            return null;
        }
        var value = node.get(field);
        return value == null || value.isNull() ? null : value.asString(null);
    }

    private static long longValue(JsonNode node, String field) {
        if (node == null || node.get(field) == null) {
            return 0;
        }
        return node.get(field).asLong(0);
    }

    private static String firstNonNull(String first, String second) {
        return first == null ? second : first;
    }

    private record User(String name, String username, String avatarUrl) {
        private static final User EMPTY = new User(null, null, null);
    }
}

File: android/app/src/main/java/dev/nytweetdeck/android/xapi/DirectMessageResponseParser.kt
package dev.nytweetdeck.android.xapi

import dev.nytweetdeck.android.model.DirectMessage
import dev.nytweetdeck.android.model.DirectMessagePage
import java.util.LinkedHashMap
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull

/** X Web DM受信箱のentries/users応答をAndroid共通モデルへ変換する。 */
class DirectMessageResponseParser(
    private val json: Json = Json,
) {
    fun parse(body: String): DirectMessagePage = try {
        val root = json.parseToJsonElement(body)
        val state = root.objectValue("inbox_initial_state") ?: root
        val users = parseUsers(state.objectValue("users"))
        val messages = state.arrayValue("entries")
            ?.mapNotNull { parseMessage(it, users) }
            ?.sortedByDescending(DirectMessage::timestamp)
            .orEmpty()

        DirectMessagePage(
            messages = messages,
            nextCursor = firstNonNull(state.text("cursor"), root.text("cursor")),
        )
    } catch (exception: XApiException) {
        throw exception
    } catch (exception: Exception) {
        throw XApiException("ダイレクトメッセージ応答を解析できません。", cause = exception)
    }

    private fun parseUsers(usersNode: JsonObject?): Map<String, User> {
        val users = LinkedHashMap<String, User>()
        usersNode?.forEach { (id, value) ->
            users[id] = User(
                name = value.text("name"),
                username = value.text("screen_name"),
                avatarUrl = value.text("profile_image_url_https"),
            )
        }
        return users
    }

    private fun parseMessage(
        entry: JsonElement,
        users: Map<String, User>,
    ): DirectMessage? {
        val message = entry.objectValue("message") ?: return null
        val data = message.objectValue("message_data")
        val id = message.text("id") ?: return null
        val senderId = data.text("sender_id") ?: return null
        val text = data.text("text") ?: return null
        val sender = users[senderId]
        return DirectMessage(
            id = id,
            conversationId = message.text("conversation_id"),
            senderId = senderId,
            senderName = sender?.name,
            senderUsername = sender?.username,
            senderAvatarUrl = sender?.avatarUrl,
            text = text,
            timestamp = message.longValue("time"),
        )
    }

    private data class User(
        val name: String?,
        val username: String?,
        val avatarUrl: String?,
    )

    private fun JsonElement?.objectValue(name: String): JsonObject? =
        (this as? JsonObject)?.get(name) as? JsonObject

    private fun JsonElement?.arrayValue(name: String): JsonArray? =
        (this as? JsonObject)?.get(name) as? JsonArray

    private fun JsonElement?.text(name: String): String? =
        ((this as? JsonObject)?.get(name) as? JsonPrimitive)?.contentOrNull

    private fun JsonElement?.longValue(name: String): Long =
        ((this as? JsonObject)?.get(name) as? JsonPrimitive)?.longOrNull ?: 0L

    private fun firstNonNull(first: String?, second: String?): String? = first ?: second
}
